dask/dataframe/core.py

```diff
@@ -4356,26 +4356,14 @@ def sort_values(
         """
         from .shuffle import sort_values
 
-        sort_kwargs = {
-            "by": by,
-            "ascending": ascending,
-            "na_position": na_position,
-        }
-        if sort_function is None:
-            sort_function = M.sort_values
-        if sort_function_kwargs is not None:
-            sort_kwargs.update(sort_function_kwargs)
-
-        if self.npartitions == 1:
-            return self.map_partitions(sort_function, **sort_kwargs)
```

Comment:
Noting that we would also need to move the single partition sort case to the internal `sort_values` so that custom sorting functions are supported there; this has the side effect of making it so multi-column sorting is no longer supported for single partition cases, which is niche enough that I don't _think_ this would cause any downstream breakage

Reply:
Could you explain this a bit more, maybe add a snippet or a test demonstrating the behavior change? From reading your comment, it seems that now the test you included in #8345 would fail, but it seems to pass?

Reply:
Sure! So before if we were to try something like

```python
import pandas as pd
import dask.dataframe as dd

df = pd.DataFrame({"x": [1, 2, 3, 1, 2, 3], "y": [1, 2, 3, 4, 5, 6]})
ddf = dd.from_pandas(df, npartitions=1)

ddf.sort_values(by=["x", "y"]).compute()
```

Because the dataframe has a single partition, we would reach the `map_partitions` call in I removed above _before_ reaching Dask's input validation checks in `shuffle.sort_values`, one of which checks that `by` is only a single column:

https://github.com/dask/dask/blob/3e7d1d03ca40972af7f8fc5735e1279b9740dab2/dask/dataframe/shuffle.py#L88-L97

Now that the `map_partitions` call is moved to be after these input validation checks, the above snippet would trigger the `NotImplementedError`, even though it is possible for the single-partition dataframe to be sorted by multiple columns.

As I'm writing this out, I realize that it should be possible to modify this check of `by` somewhat to allow multiple columns when the dataframe has a single partition - I will try making those changes now 🙂 

> it seems that now the test you included in #8345 would fail, but it seems to pass?

That test achieves multi-column sorting in a different, somewhat hacky way - by passing a single by-column to Dask's `sort_values` and multiple by-columns to the custom sorting function (which is essentially just a wrapper for the partition library's `sort_values`), Dask will perform a rough initial "sort" using the single column it has been provided before passing off the remaining work of the multi-column sort to the partition library (pandas, cuDF, etc.) that actually supports multi-column sorting.

Reply:
Yeah that's a good idea - I'll add a single partition sorting test